src/sys/windows/tcp.rs

```diff
@@ -392,12 +400,26 @@ impl StreamImp {
         me.iocp.set_readiness(me.iocp.readiness() & !Ready::writable());
 
         trace!("scheduling a write");
-        let err = unsafe {
+        let ret = unsafe {
             let overlapped = miow::Overlapped::from_raw(self.inner.write.as_mut_ptr());
-            self.inner.socket.write_overlapped(&buf[pos..], overlapped)
+            match self.inner.socket.write_overlapped(&buf[pos..], overlapped) {
+                Ok(true) if me.instant_notify => {
+                    self.inner.socket.result(overlapped).map(Some)
+                }
+                Ok(_) => Ok(None),
+                Err(e) => Err(e),
+            }
         };
-        match err {
-            Ok(_) => {
+        match ret {
+            Ok(Some((transfered_bytes, _))) => {
+                me.write = if transfered_bytes == buf.len() - pos {
+                    self.add_readiness(me, Ready::writable());
+                    State::Empty
+                } else {
+                    State::Pending((buf, pos + transfered_bytes))
```

Comment:
I also have my doubts that this is actually a valid case but in theory yes.
I'll add the loop if we have some conclusive benchmark results for the `wsagetoverlappedresult` stuff (reducing the commit spam a bit by combining those changes, if the benchmarks show they make sense ).